compiler/src/main/kotlin/io/github/apl_cornell/viaduct/algebra/LatticeCongruence.kt

```diff
@@ -0,0 +1,62 @@
+package io.github.apl_cornell.viaduct.algebra
+
+typealias Congruence<A> = Pair<A, A>
+
+interface LatticeCongruence<A : Lattice<A>> {
+    fun equals(first: A, second: A): Boolean
+    fun lessThanOrEqualTo(first: A, second: A): Boolean
+}
+
+class FreeDistributiveLatticeCongruence<A>(
+    private val congruence: List<Congruence<FreeDistributiveLattice<A>>>
+) : LatticeCongruence<FreeDistributiveLattice<A>> {
+
+    private val foldedCongruence: Congruence<FreeDistributiveLattice<A>> =
+        congruence.fold(
+            Pair(
+                FreeDistributiveLattice.bounds<A>().bottom,
+                FreeDistributiveLattice.bounds<A>().top
+            )
+        ) { acc, element ->
+            if (element.first.meet(acc.second) == element.first) {
+                Pair(acc.first.meet(element.first), acc.second.join(element.second))
+            } else {
+                assert(element.first.meet(acc.second) == element.second)
+                Pair(acc.first.meet(element.second), acc.second.join(element.first))
+            }
+        }
+
+    override fun equals(first: FreeDistributiveLattice<A>, second: FreeDistributiveLattice<A>) =
```

Comment:
I don't think this is a correct implementation. Didn't you show folded congruences algorithm doesn't work?